### PR TITLE
Add encoding and writing JPEG ops

### LINKED ISSUE #2691: Reading PNG/JPG images into a torch::tensor and saving a torch::tensor to PNG/JPG in C++ without OpenCV
## 🚀 Feature

After integrating Siv3D with Libtorch (https://github.com/QuantScientist/Siv3DTorch) I am now trying to read and write images from and to Siv3D **in C++, not Python**. The way it works is:

## Motivation
In C++ I need to do the following:
1. An image is read from disk (usually using OpenCV which is easy but I am trying to avoid)
2. The image is converted to torch::tensor
3. A DL model is applied on the tensor
4. A tensor is returned from the model
5. The tensor is converted to an image for display/saving purposes.

This is one example where they used stb_image to this, avoiding the use of OpenCV.
https://github.com/prabhuomkar/pytorch-cpp/blob/master/utils/image_io/src/image_io.cpp

## Alternatives

For reference this is the OpenCV to Libtorch conversion utils which I use, I would like something very similiar:
```
at::Tensor matToTensor(cv::Mat frame, int h, int w, int c) {
    cv::cvtColor(frame, frame, CV_BGR2RGB);
    frame.convertTo(frame, CV_32FC3, 1.0f / 255.0f);
    auto input_tensor = torch::from_blob(frame.data, {1, h, w, c});
    input_tensor = input_tensor.permute({0, 3, 1, 2});

    torch::DeviceType device_type = torch::kCPU;
//    if (torch::cuda::is_available()) {
    device_type = torch::kCUDA;
//    }
    input_tensor = input_tensor.to(device_type);
    return input_tensor;
}

cv::Mat tensorToOpenCv(at::Tensor out_tensor, int h, int w, int c) {
    out_tensor = out_tensor.squeeze().detach().permute({1, 2, 0});
    out_tensor = out_tensor.mul(255).clamp(0, 255).to(torch::kU8);
    out_tensor = out_tensor.to(torch::kCPU);
    cv::Mat resultImg(h, w, CV_8UC3);
    // cv::Mat resultImg(h, w, CV_8UC1);
    std::memcpy((void *) resultImg.data, out_tensor.data_ptr(), sizeof(torch::kU8) * out_tensor.numel());
    return resultImg;
}

```
## Additional context

I found this:"https://github.com/pytorch/vision/blob/5e4a9f6d1a2bf85137f4826dbf76e4f25986f878/torchvision/csrc/cpu/image/readpng_cpu.cpp
however, could not get any useful method out of it. 

Thanks,

### PR DESCRIPTION
**Pull request by andfoy**
This PR introduces new encoding and writing operators to the torchvision image frontend, as follows:

- [x] `encode_jpeg`: Takes a CHW tensor as input and produces a byte buffer containing the JPEG file in tensor form.
- [x] `write_jpeg`: Takes a CHW tensor and a filename and writes the image into a JPEG file in disk.

Fixes #2691

### CODE REVIEW

**Review by fmassa — COMMENTED** [review]
This is looking good, thanks!

Can you already add tests for JPEG now, so that the ops for PNG encoding could be done in a separate PR?

File: torchvision/csrc/cpu/image/writejpeg_cpu.cpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    TORCH_CHECK(false, (const char*)jerr.jpegLastErrorMsg);
+  }
+
+  // Get image info
+  int channels, height, width;
+
+  // Move tensor to CPU and cast it into uint8
+  torch::Tensor input = data.to(torch::kCPU).to(torch::kU8);
```
Comment (fmassa) [inline torchvision/csrc/cpu/image/writejpeg_cpu.cpp:61]: Instead of casting to `uint8`, can't you instead raise an error in the C++ side as well if the tensor is not uint8?

I also wonder about silently casting a CUDA tensor to CPU, it could give the impression that we are encoding on the GPU. Shouldn't we maybe just raise an error if the input is not on the CPU?

File: torchvision/csrc/cpu/image/writejpeg_cpu.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+  TORCH_CHECK(
+      data.dim() >= 2 && data.dim() <= 3,
```
Comment (fmassa) [inline torchvision/csrc/cpu/image/writejpeg_cpu.cpp:64]: I wonder if we should support the `dim == 2` case. In one hand, it can be handy, but on the other hand, it might be preferable to enforce that all the tensors always have 3 dimensions -- for example, our decoding ops don't give the option to return 2d tensors

File: torchvision/csrc/cpu/image/writejpeg_cpu.cpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  jpeg_finish_compress(&cinfo);
+  jpeg_destroy_compress(&cinfo);
+
+  torch::TensorOptions options = torch::TensorOptions{torch::kU8};
+  auto outTensor = torch::empty({(long)jpegSize}, options);
+
+  // Copy memory from jpeg buffer, since torch cannot get ownership of it via
+  // `from_blob`
+  auto outPtr = outTensor.data_ptr<uint8_t>();
+  std::memcpy(outPtr, jpegBuf, sizeof(uint8_t) * outTensor.numel());
```
Comment (fmassa) [inline torchvision/csrc/cpu/image/writejpeg_cpu.cpp:121]: For the future: I believe it might be possible to directly use `outTensor.data_ptr<uint8_t>()` for `jpeg_write_scanlines`, as I believe TF does it via [this function](https://github.com/tensorflow/tensorflow/blob/e8598ce0454c440fca64e4ebc4aeedfa7afd5c97/tensorflow/core/lib/jpeg/jpeg_handle.cc#L95-L101) in [their Compress function](https://github.com/tensorflow/tensorflow/blob/e8598ce0454c440fca64e4ebc4aeedfa7afd5c97/tensorflow/core/lib/jpeg/jpeg_mem.cc#L656).

Can you open a new issue to track down this potential micro-optimization?

File: torchvision/csrc/cpu/image/writejpeg_cpu.cpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  auto jpegBuf = encodeJPEG(data, quality);
+  auto fileBytes = jpegBuf.data_ptr<uint8_t>();
+  auto fileCStr = filename.c_str();
+  FILE* outfile = fopen(fileCStr, "wb");
+
+  TORCH_CHECK(outfile != NULL, "Error opening output jpeg file");
+
+  fwrite(fileBytes, sizeof(uint8_t), jpegBuf.numel(), outfile);
+  fclose(outfile);
+}
```
Comment (fmassa) [inline torchvision/csrc/cpu/image/writejpeg_cpu.cpp:141]: This is fine, although I would have expected it to be left to be implemented in Python entirely (like we do with the `read_jpeg`.

**Review by fmassa — COMMENTED** [review]
Thanks a lot for the PR!

I have a few comments -- in particular, the fact that we don't need to call `contiguous()` makes me a bit suspicious.

File: torchvision/csrc/cpu/image/writejpeg_cpu.cpp [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+  TORCH_CHECK(
+      data.dim() >= 2 && data.dim() <= 3,
+      "Input data should be a 3-dimensional or a 2-dimensional tensor");
+
+  if (data.dim() == 3) {
+    channels = data.size(0);
+    height = data.size(1);
+    width = data.size(2);
+    input = input.permute({1, 2, 0});
```
[the remark this answers is not in the document]
Comment (fmassa) [reply torchvision/csrc/cpu/image/writejpeg_cpu.cpp:71]: After a second thought, I realized that this only works because in your tests your directly read from `read_jpeg`, which returns non-contiguous tensors, and when permuting it it becomes contiguous. So we do need a `.contiguous()` here to make the code more robust. Can you also add a test for a contiguous tensor?

File: torchvision/csrc/cpu/image/writejpeg_cpu.cpp (hunk 3 above)
Comment (andfoy) [reply torchvision/csrc/cpu/image/writejpeg_cpu.cpp:121]: I think that we cannot use a memory buffer that was not created by libjpeg directly: https://github.com/libjpeg-turbo/libjpeg-turbo/blob/f81833aed68ec7d04ecdeb7f4f7566040f24300e/jdatadst.c#L267

File: torchvision/csrc/cpu/image/writejpeg_cpu.cpp (hunk 4 above)
Comment (andfoy) [reply torchvision/csrc/cpu/image/writejpeg_cpu.cpp:141]: I can move the save part to Python, however I was thinking this function to be exposed to the C++ API

File: torchvision/csrc/cpu/image/writejpeg_cpu.cpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+  int height = data.size(1);
+  int width = data.size(2);
+  auto input = data.permute({1, 2, 0}).contiguous();
+
+  std::ostringstream channelErrS;
+  channelErrS << "The number of channels should be 1 or 3, got: " << channels;
+
+  const std::string& channelErrStr = channelErrS.str();
+  const char* channelErr = channelErrStr.c_str();
+  TORCH_CHECK(channels == 1 || channels == 3, channelErr);
```
Comment (fmassa) [inline torchvision/csrc/cpu/image/writejpeg_cpu.cpp:71]:
```suggestion
  TORCH_CHECK(channels == 1 || channels == 3, "The number of channels should be 1 or 3, got: ", channels);
```

File: torchvision/io/image.py [hunk 7]
```diff
@@ -102,3 +102,48 @@ def read_jpeg(path: str) -> torch.Tensor:
...
+        quality (int): Quality of the resulting JPEG file, it must be a number
+    between 1 and 100. Default: 75
+    """
+    if quality < 1 or quality > 100:
+        raise ValueError('Image quality should be a positive number '
+                         'between 1 and 100')
+
+    if not input.dtype == torch.uint8:
+        raise ValueError("Expected a torch.uint8 tensor.")
+
```
Comment (fmassa) [inline torchvision/io/image.py:148]: nit: this can now be removed as it's checked by the C++ code
```suggestion
```

File: torchvision/io/image.py [hunk 8]
```diff
@@ -102,3 +102,48 @@ def read_jpeg(path: str) -> torch.Tensor:
...
+        output (Tensor[1]): A one dimensional int8 tensor that contains the raw
+    bytes of the JPEG file.
+    """
+    if quality < 1 or quality > 100:
+        raise ValueError('Image quality should be a positive number '
+                         'between 1 and 100')
+
+    if not input.dtype == torch.uint8:
+        raise ValueError("Expected a torch.uint8 tensor.")
+
```
Comment (fmassa) [inline torchvision/io/image.py:126]: nit: this can now be removed as it's checked by the C++ code

```suggestion
```

File: torchvision/csrc/cpu/image/jpegcommon.h [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+#include <cstdio>
+#include <cstddef>
+// clang-format on
+#include <jpeglib.h>
+#include <setjmp.h>
+#include <string>
+
+static const JOCTET EOI_BUFFER[1] = {JPEG_EOI};
+static const size_t JPEG_BUF_SIZE = 16384;
+
```
Comment (fmassa) [inline torchvision/csrc/cpu/image/jpegcommon.h:14]: nit: this doesn't seem to be used
```suggestion
```

File: torchvision/csrc/cpu/image/jpegcommon.h [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+#pragma once
```
Comment (fmassa) [inline torchvision/csrc/cpu/image/jpegcommon.h:2]:
```suggestion
#pragma once
```

File: torchvision/csrc/cpu/image/jpegcommon.cpp [hunk 11]
```diff
@@ -0,0 +1,18 @@
+
+#include "jpegcommon.h"
```
Comment (fmassa) [inline torchvision/csrc/cpu/image/jpegcommon.cpp:2]:
```suggestion
#include "jpegcommon.h"
```

File: test/test_image.py [hunk 12]
```diff
@@ -66,6 +68,65 @@ def test_damaged_images(self):
...
+            expected_file = os.path.join(
+                write_folder, '{0}_pil.jpg'.format(filename))
+            img = read_jpeg(img_path)
+
+            # PIL sets jpeg quality to 75 by default
+            jpeg_bytes = encode_jpeg(img, quality=75)
+            with open(expected_file, 'rb') as f:
+                pil_bytes = f.read()
+                pil_bytes = torch.as_tensor(list(pil_bytes), dtype=torch.uint8)
+            self.assertTrue(jpeg_bytes.equal(pil_bytes))
```
Comment (fmassa) [inline test/test_image.py:85]: I propose we test for contiguous and non-contiguous tensors as well
```suggestion
            with open(expected_file, 'rb') as f:
                pil_bytes = f.read()
                pil_bytes = torch.as_tensor(list(pil_bytes), dtype=torch.uint8)
                
            for src_img in [img, img.contiguous()]:
                # PIL sets jpeg quality to 75 by default
                jpeg_bytes = encode_jpeg(src_img, quality=75)
                self.assertTrue(jpeg_bytes.equal(pil_bytes))
```

File: torchvision/csrc/cpu/image/writejpeg_cpu.cpp (hunk 3 above)
Comment (fmassa) [reply torchvision/csrc/cpu/image/writejpeg_cpu.cpp:109]: that's fine if this suggestion doesn't work.

File: torchvision/csrc/cpu/image/writejpeg_cpu.cpp (hunk 4 above)
Comment (fmassa) [reply torchvision/csrc/cpu/image/writejpeg_cpu.cpp:129]: Let's keep this as is for now, this was just a suggestion. We could revisit this in the future.

**Review by fmassa — COMMENTED** [review]
Thanks a lot, this looks good to me!

I made a few inline suggestions.
Also, I believe we can remove `test/assets/jpeg_write/grace_hopper_517x606_pil.pth` prior to commiting, as it doesn't seem like it's used anymore

File: torchvision/csrc/cpu/image/writejpeg_cpu.cpp [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+  // Check that the input tensor is 3-dimensional
+  TORCH_CHECK(data.dim() == 3, "Input data should be a 3-dimensional tensor");
+
+  // Get image info
+  int channels = data.size(0);
+  int height = data.size(1);
+  int width = data.size(2);
+  auto input = data.permute({1, 2, 0}).contiguous();
+
+  TORCH_CHECK(channels == 1 || channels == 3, "The number of channels should be 1 or 3, got: ", channels);
```
Comment (fmassa) [inline torchvision/csrc/cpu/image/writejpeg_cpu.cpp:66]:
```suggestion
  TORCH_CHECK(
      channels == 1 || channels == 3,
      "The number of channels should be 1 or 3, got: ",
      channels);
```

File: test/test_image.py [hunk 14]
```diff
@@ -66,6 +68,67 @@ def test_damaged_images(self):
...
+            filename, _ = os.path.splitext(os.path.basename(img_path))
+            write_folder = os.path.join(dirname, 'jpeg_write')
+            expected_file = os.path.join(
+                write_folder, '{0}_pil.jpg'.format(filename))
+            img = read_jpeg(img_path)
+
+            with open(expected_file, 'rb') as f:
+                pil_bytes = f.read()
+                pil_bytes = torch.as_tensor(list(pil_bytes), dtype=torch.uint8)
+
```
Comment (fmassa) [inline test/test_image.py:83]:
```suggestion

```

**Review by fmassa — APPROVED** [review]
Thanks a lot Edgar!

**Comment by faranaziz** [comment]
How to I access this code from C++? I mean, I could not find it inside libtorch-win-shared-with-deps-1.8.1+cpu.zip. Is it in another library?

**Comment by fmassa** [comment]
Hi @faranaziz 

We haven't exposed the IO ops in the same binaries as the other modeling ops in torchvision. If you want those ops, you'll need to compile torchvision from source
